src/OmniSharp.MSBuild/ProjectManager.cs

```diff
@@ -406,38 +398,50 @@ private void WatchProjectFiles(ProjectFileInfo projectFileInfo)
             {
                 _fileSystemWatcher.Watch(projectFileInfo.RuleSet.FilePath, (file, changeType) =>
                 {
-                    QueueProjectUpdate(projectFileInfo.FilePath, allowAutoRestore: false, projectFileInfo.ProjectIdInfo);
+                    QueueProjectUpdate(projectFileInfo.FilePath, allowAutoRestore: false, projectFileInfo.ProjectIdInfo, changeType);
                 });
             }
 
-            if (!string.IsNullOrEmpty(projectFileInfo.ProjectAssetsFile))
+            foreach(var projectAssetFile in GetProjectAssetFilePaths(projectFileInfo))
             {
-                _fileSystemWatcher.Watch(projectFileInfo.ProjectAssetsFile, (file, changeType) =>
+                _fileSystemWatcher.Watch(projectAssetFile, (file, changeType) =>
                 {
-                    QueueProjectUpdate(projectFileInfo.FilePath, allowAutoRestore: false, projectFileInfo.ProjectIdInfo);
+                    QueueProjectUpdate(projectFileInfo.FilePath, allowAutoRestore: false, projectFileInfo.ProjectIdInfo, changeType);
                 });
+            }
+        }
 
-                var restoreDirectory = Path.GetDirectoryName(projectFileInfo.ProjectAssetsFile);
-                var nugetFileBase = Path.Combine(restoreDirectory, Path.GetFileName(projectFileInfo.FilePath) + ".nuget");
-                var nugetCacheFile = nugetFileBase + ".cache";
-                var nugetPropsFile = nugetFileBase + ".g.props";
-                var nugetTargetsFile = nugetFileBase + ".g.targets";
+        private void UnwatchProjectFiles(ProjectFileInfo projectFileInfo)
+        {
+            _fileSystemWatcher.Unwatch(projectFileInfo.FilePath);
 
-                _fileSystemWatcher.Watch(nugetCacheFile, (file, changeType) =>
-                {
-                    QueueProjectUpdate(projectFileInfo.FilePath, allowAutoRestore: false, projectFileInfo.ProjectIdInfo);
-                });
+            if (projectFileInfo.RuleSet?.FilePath != null)
+            {
+                _fileSystemWatcher.Unwatch(projectFileInfo.RuleSet.FilePath);
```

Comment:
Will this be problematic if multiple projects use the same ruleset file? `project.AnalyzerConfigDocuments` are another group of files that could be unwatched, but they are likely shared with other projects so it may require reference counting.